### DISCUSSION TITLE
half-width character display bug！

### DISCUSSION THREAD

**Discussion by heyhumor** [pymupdf/PyMuPDF#1248] [Looking for help] [answered]
## Describe the bug 
Recently, I find a bug with function [ Page.add_freetext_annot() ]. 
When i put text with half-width character, just like '9999ｷ', '9999ｳ' or '9999ﾀ', the number '9999' will display with full-with mode.
But when i double click with this freetext annotation , It's back to normal, confused a lot.

## Code
page.add_freetext_annot(rect=fitz.Rect(353, 687, 365, 750), text='9999ﾀ', align=1, rotate=90,
                                        fontname='HeIt', fontsize=12, fill_color=(1, 0.75, 0.80))
![2021-09-09_083314](https://user-images.githubusercontent.com/52818042/132603251-30b92671-f730-4437-8ff5-e0a35e41be18.png)

**Comment by heyhumor** [comment]
Sorry i add one thing, the character 'ﾀ' above is Janpanese, And this bug is not with fonttype, i tried all fonttype.
I just need the number display normal, please give me some advice , thanks!

**Comment by JorjMcKie** [comment, marked as answer]
This is not a bug:
For FreeText annots, MuPDF supports non-embedded fonts only: Base-14 and CJK. As explained [here](https://pymupdf.readthedocs.io/en/latest/faq.html#changing-annotations-unexpected-behaviour), surprising behaviour can occur if you access / update annotations with a different tool than the one you used for creating them.
As you wrote, you "double clicked" the annot. This will have happened in a PDF viewer different from MuPDF (maybe Adobe Acrobat or something else), so this effect is inevitable.
MuPDF's internal logic for creating the annotation works roughly like this:
* it tries to minimize the fonts used overall
* if encountering Japanese, it will add non-embeddable font "Mincho" to the annot, similar is true for Korean and Chinese
* if no character is unsupported by one of the CJK fonts, it will not include "helv" - even if you specify it
* if your annot text starts with some char unsupported by CJK, "helv" will be used as a start and only be switched to a CJK font when required. After this it will stay with the current font as long as possible, etc.

Look at this snippet to see this work:
```python
import fitz

doc = fitz.open()
page = doc.new_page()
r1 = fitz.Rect(100, 100, 300, 120)
r2 = fitz.Rect(100, 120, 300, 140)
t1 = "9999ﾀ"  # all chars are supported by font "Mincho" (Japanese)
t2 = "x9999ﾀ"  # starts with a char unsupported by Mincho
page.add_freetext_annot(r1, t1, fontname="helv")
page.add_freetext_annot(r2, t2, fontname="helv")
doc.ez_save(__file__.replace(".py", ".pdf"), pretty=True)
```
This will produce the following:
![grafik](https://user-images.githubusercontent.com/8290722/132633574-ed1e3348-b8d0-487e-929f-c55dcbc81829.png)

When looking at the PDF "source" of the annotations, we will see this:
```postscript
>>> a1 = page.first_annot
>>> a2 = a1.next
>>> print(a1._getAP().decode())
0 w
0 0 200 20 re
W
n
BT
0 0 0 rg
0 24.8 Td
0 -14.400001 Td
/Mincho 12 Tf
<0039003900390039ff80> Tj  % ==> all text in Mincho!
ET

>>> print(a2._getAP().decode())
0 w
0 0 200 20 re
W
n
BT
0 0 0 rg
0 24.8 Td
0 -14.400001 Td
/Helv 12 Tf
(x9999) Tj  % ==> start text in Helv
/Mincho 12 Tf
<ff80> Tj  % ==> switch to Mincho when we must
ET

>>> 
```

**Comment by heyhumor** [comment]
Got it, Thanks for your anwser！
But also it's a trouble to me, i have lots of labels like above. So, have any function could edit with freetext annotations.
Then i can put '9999' first and append with 'ﾀ' in annotations, what's your idea?

**Reply by JorjMcKie** [reply]
That's not an easy one and is guaranteed to be hacky.
1. First of all you must make sure that the `/Resources` of the associated `AP/N` object contains your chosen Base-14 font.
2. Second you must create your own version of the PDF "source" that I mentioned above - this is the hard part.

Re 1.
Get the annot xref and read out the `AP/N` xref:
```python
>>> doc.xref_get_key(annot.xref, "AP/N")
('xref', '9 0 R')
>>> apnxref = doc.xref_get_key(annot.xref, "AP/N")[1].split()[0]
>>> apnxref = int(apnxref)
>>> apnxref
9
```
Inside the APN object add the Helvetica font (assuming this is your non-CJK font):
```python
>>> print(doc.xref_object(9))
<<
  /Type /XObject
  /Subtype /Form
  /BBox [ 0 0 200 20 ]
  /Matrix [ 1 0 0 1 0 0 ]
  /Resources <<
    /Font <<
      /Mincho 6 0 R
    >>
  >>
  /Length 87
  /Filter /FlateDecode
>>
>>> doc.xref_get_key(9, "Resources/Font")
('dict', '<</Mincho 6 0 R>>')
>>> fontdict = doc.xref_get_key(9, "Resources/Font")[1]
>>> # add /Helv font
>>> fontdict = fontdict.replace(">>", "/Helv 4711 0 R>>")
>>> # 4711 stands for the /Helv font xref: to be determined!
>>> doc.xref_set_key(9, "Resources/Font", fontdict)
>>> print(doc.xref_object(9))  # confirm we did well:
<<
  /Type /XObject
  /Subtype /Form
  /BBox [ 0 0 200 20 ]
  /Matrix [ 1 0 0 1 0 0 ]
  /Resources <<
    /Font <<
      /Mincho 6 0 R
      /Helv 4711 0 R
    >>
  >>
  /Length 87
  /Filter /FlateDecode
>>
>>> 
```
If you do not know the Helvetica xref or do not have used that font yet, you need to insert it: `xref = page.insert_font()`.

Re 2.
To modify the PDF "source" of the annot, you can take the example above:
```python
>>> print(a1._getAP().decode())
0 w
0 0 200 20 re
W
n
BT
0 0 0 rg
0 24.8 Td
0 -14.400001 Td
/Mincho 12 Tf
<0039003900390039ff80> Tj  % ==> all text in Mincho!
ET
```
What we see here: "0039" = hex(ord("9")). And 0xFF80 = hex(ord("ﾀ")). So you would have to replace the two lines
```
/Mincho 12 Tf
<0039003900390039ff80>
```
by these 4 lines:
```
/Helv 12 Tf
(9999) Tj
/Mincho 12 Tf
<ff80> Tj
```
Once you have finished that string (call it "newsource"), put it back to the PDF:
```python
annot._setAP(newstring.encode())  # updates the annot source
```

**Reply by JorjMcKie** [reply]
As per point 2, you could of course also just insert
```python
/Mincho 12 Tf  % "12" means the fontsize!
<ff80> Tj
```
before the ``ET`` string after you have created your annot with Helvetica text. But still you need to make sure, that the APN object's `/Resources/Font` references the Mincho font.
Similar to above you can create it via ``xref = page.insert_font(fontname="japan-s")``.

**Reply by heyhumor** [reply]
Got it, I will try it later ! 
Thanks so much for you answers,  sincerely.

**Reply by JorjMcKie** [reply]
Try this script - it works for me:
```python
import fitz

fontsize=12
doc = fitz.open()
page = doc.new_page()
r1 = fitz.Rect(100, 100, 300, 120)
t1 = "9999"
append_this = b"/Mincho %g Tf\n<FF80> Tj\nET" % fontsize
annot = page.add_freetext_annot(r1, t1, fontname="helv", fontsize=fontsize)
apnxref = doc.xref_get_key(annot.xref, "AP/N")[1]
apnxref = int(apnxref.replace("0 R", ""))
mincho_xref = page.insert_font(fontname="japan-s")
fontdict = doc.xref_get_key(apnxref, "Resources/Font")[1]
fontdict = fontdict.replace(">>", "/Mincho %i 0 R>>" % mincho_xref)
doc.xref_set_key(apnxref, "Resources/Font", fontdict)
print(doc.xref_object(apnxref))
source = annot._getAP()
source = source.replace(b"ET", append_this)
annot._setAP(source)
print(source.decode())
doc.ez_save(__file__.replace(".py", ".pdf"), pretty=True)
```

**Reply by JorjMcKie** [reply]
With the last code snippet there is one remaining little issue:
It will create the annotation like this:
```pdf
4 0 obj
<<
  /Type /Annot
  /Subtype /FreeText
  /Rect [ 100 722 300 742 ]
  /BS <<
    /W 0
  >>
  /DA (0 0 0 0 k /Helv 12 Tf)
  /P 3 0 R
  /F 4
  /Contents (9999)
  /Rotate 0
  /Q 0
  /NM (fitz-A0)
  /AP <<
    /N 6 0 R
  >>
>>
endobj
```
Strictly speaking, the entry `/Contents (9999)` is not correct, because it does not equal the full final text "9999ﾀ". This is no problem at all ... as long as the annotation is not updated, either with PyMuPDF or any other software.
But you can update the annot itself with the same method as shown above:
``doc.xref_set_key(annot.xref, "Contents", fitz.get_pdf_str("9999ﾀ"))``.
If you update with PyMuPDF, the same unpleasant layout will be recreated, so you need to apply the hack again.

**Reply by heyhumor** [reply]
Hey, Friend, Thanks for your code first.
I have tried all ways you provided. And the way of change "AP/N" source with /Helv font  is worked. 
But it seems every pdf's worker has theirown handler. It work on Adobe Acrobat Pro DC, But the number on PDF viewer it can't shown...
![Uploading 2021-09-13_113937.png…]()
And i am sorry to tell you, i can't modify their real value, i need to read and deal with them, so "/Contents (9999)"  is useless, Even it worked.

**Reply by heyhumor** [reply]
![Uploading 2021-09-13_113937.png…]()
